File: DIP/DIP_06/Python/DIP_006_Sobel_Operator.py
import cv2
import numpy as np

image = cv2.imread('dragon.jpg',0)
cv2.imshow('Original image',image)
cv2.waitKey()

# We have already explained linear filters for horizontal and vertical edge detection

# this is how we can create a horizontal edge detector.
# cv2.Sobel(src_gray, depth, x_order, y_order)
# src_gray: The input image.
# depth: The depth of the output image.
# x_order: The order of the derivative in x direction.
# y_order: The order of the derivative in y direction.
# To calculate the gradient in x direction we use: x_order= 1 and y_order= 0.
# To calculate the gradient in x direction we use: x_order= 0 and y_order= 1.
image_X = cv2.Sobel(image, cv2.CV_8UC1, 1, 0)

cv2.imshow("Sobel image", image_X)
cv2.waitKey()

image_Y = cv2.Sobel(image, cv2.CV_8UC1, 0, 1)
cv2.imshow("Sobel image", image_Y)
cv2.waitKey()

# When we combine the horizontal and vertical edge detector together
sobel = cv2.add(image_X, image_Y)
cv2.imshow("Sobel - L1 norm", sobel)
cv2.waitKey()

(_, sobmin, _, sobmax) = cv2.minMaxLoc(sobel)
# this idea is inspired from the book
# "Robert Laganiere Learning OpenCV 3:: computer vision"
# what it actually does, makes the non-edges to white values
# and edges to dark values, so that it is more common for our visual interpretation.
# this is done according to formula
# sobelImage = - alpha * sobel +  255;

#sobel.convertTo(sobelImage, CV_8UC1, -255./sobmax, 255);

sobelImage = np.int16(sobel)     # convert to signed 16 bit integer to allow overflow
sobelImage = np.clip(sobelImage, -255./sobmax[0], 255) # force all values to be between -255./sobmax and 255

# after clip img2 is effectively unsigned 8 bit, but make it explicit:
sobelImage = np.uint8(sobelImage)

cv2.imshow("Edges with a sobel detector", sobelImage)
cv2.waitKey()

(_, sobmin, _, sobmax) = cv2.minMaxLoc(sobelImage)
# image_Laplacian = image_Laplacian / max_value * 255

image_Sobel_thresholded = cv2.threshold(sobelImage, 20, 255, cv2.THRESH_BINARY)
cv2.imshow("Thresholded Laplacian", image_Sobel_thresholded[1])
cv2.waitKey()

# Also, very popular filter for edge detection is Laplacian operator
# It calculates differences in both x and y direction and then sums their amplitudes.

# here we will apply low pass filtering in order to better detect edges
# try to uncomment this line and the result will be much poorer.
image = cv2.GaussianBlur(image,(5,5),1)

image_Laplacian = cv2.Laplacian(image, cv2.CV_8UC1)

cv2.imshow("The Laplacian", image_Laplacian)
cv2.waitKey()

(_, min_value1, _, max_value1) = cv2.minMaxLoc(image_Laplacian)
#image_Laplacian = image_Laplacian / max_value * 255;

image_Laplacian_thresholded = cv2.threshold(image_Laplacian,70, 220, cv2.THRESH_BINARY)
cv2.imshow("Thresholded Laplacian", image_Laplacian_thresholded[1])
cv2.waitKey()

File: DIP/DIP_06/C++/DIP_006_Sobel_Operator.cpp
#include <iostream>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

int main()
{

    Mat image = cv::imread("dragon.jpg", IMREAD_GRAYSCALE);
    cv::imshow("Dragon", image);
    cv::waitKey();

    // we have already explained linear filters for horizontal and vertical edge detection
    // reference -->> CNN #002, CNN #003 posts.

    cv::Mat image_X;
    // this is how we can create a horizontal edge detector
    // Sobel(src_gray, dst, depth, x_order, y_order)
    // src_gray: The input image.
    // dst: The output image.
    // depth: The depth of the output image.
    // x_order: The order of the derivative in x direction.
    // y_order: The order of the derivative in y direction.
    // To calculate the gradient in x direction we use: x_order= 1 and y_order= 0.
    // To calculate the gradient in x direction we use: x_order= 0 and y_order= 1.
    cv::Sobel(image, image_X, CV_8UC1, 1, 0);
    cv::imshow("Sobel image", image_X);
    cv::waitKey();

    cv::Mat image_Y;
    // this is how we can create a vertical edge detector.
    cv::Sobel(image, image_Y, CV_8UC1, 0, 1);
    cv::imshow("Sobel image", image_Y);
    cv::waitKey();

    // When we combine the horizontal and vertical edge detector together
    cv::Mat sobel = image_X + image_Y;
    cv::imshow("Sobel - L1 norm", sobel);
    cv::waitKey();

    // this idea is inspired from the book
    // "Robert Laganiere Learning OpenCV 3:: computer vision"
    // what it actually does, makes the non-edges to white values
    // and edges to dark values, so that it is more common for our visual interpretation.
    // this is done according to formula
    // sobelImage = - alpha * sobel +  255;
    double sobmin, sobmax;
    cv::minMaxLoc(sobel, &sobmin, &sobmax);

    cv::Mat sobelImage;
    sobel.convertTo(sobelImage, CV_8UC1, -255./sobmax, 255);
    cv::imshow("Edges with a sobel detector", sobelImage);
    cv::waitKey();


    cv::Mat image_Sobel_thresholded;
    double max_value, min_value;
    cv::minMaxLoc(sobelImage, &min_value, &max_value);
    //image_Laplacian = image_Laplacian / max_value * 255;


    cv::threshold(sobelImage, image_Sobel_thresholded, 20, 255, cv::THRESH_BINARY);
    cv::imshow("Thresholded Sobel", image_Sobel_thresholded);
    cv::waitKey();
    // Also, very popular filter for edge detection is Laplacian operator
    // It calculates differences in both x and y direction and then sums their amplitudes.

    cv::Mat image_Laplacian;
    // here we will apply low pass filtering in order to better detect edges
    // try to uncomment this line and the result will be much poorer.
    cv::GaussianBlur(image, image, Size(5,5), 1);

    cv::Laplacian(image, image_Laplacian, CV_8UC1);
    cv::imshow("The Laplacian", image_Laplacian);
    cv::waitKey();

    cv::Mat image_Laplacian_thresholded;
    double max_value1, min_value1;
    cv::minMaxLoc(image_Laplacian, &min_value1, &max_value1);
    //image_Laplacian = image_Laplacian / max_value * 255;


    cv::threshold(sobel, image_Laplacian_thresholded, 70, 220, cv::THRESH_BINARY);
    cv::imshow("Thresholded Laplacian", image_Laplacian_thresholded);
    cv::waitKey();

    return 0;
}
